presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionaryStreamReader.java

```diff
@@ -154,72 +152,56 @@ public Block readBlock(Type type)
             }
         }
 
-        assureVectorSize();
-
-        // idsVector will be filled sub-batch by sub-batch.
         int[] idsVector = new int[nextBatchSize];
-        int idsVectorOffset = 0;
-        while (nextBatchSize > idsVectorOffset) {
-            int subBatchSize = min(nextBatchSize - idsVectorOffset, MAX_BATCH_SIZE);
-            if (presentStream == null) {
-                if (dataStream == null) {
-                    throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
-                }
-                Arrays.fill(isNullVector, false);
-                dataStream.nextIntVector(subBatchSize, idsVector, idsVectorOffset);
+        if (presentStream == null) {
+            if (dataStream == null) {
+                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+            }
+            if (inDictionaryStream == null) {
+                dataStream.nextIntVector(nextBatchSize, idsVector, 0);
             }
             else {
-                int nullValues = presentStream.getUnsetBits(subBatchSize, isNullVector);
-                if (nullValues != subBatchSize) {
-                    if (dataStream == null) {
-                        throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+                for (int i = 0; i < nextBatchSize; i++) {
+                    idsVector[i] = toIntExact(dataStream.next());
+                    if (!inDictionaryStream.nextBit()) {
+                        // row group dictionary elements are after the main dictionary
+                        idsVector[i] += stripeDictionarySize;
                     }
-                    dataStream.nextIntVector(subBatchSize, idsVector, idsVectorOffset, isNullVector);
                 }
             }
-
-            if (inDictionaryStream == null) {
-                Arrays.fill(inDictionaryVector, true);
+        }
+        else {
+            if (dataStream == null) {
+                // Just check if it's all null values
```

Comment:
I believe this works because `null` is ID 0.  If that assumption is true, can you add a comment?

Reply:
Yes in that case the id should be 0 because the only element in the dictionary is null and dictionaryBlock.getPositionCount() is 1.